2023/adv2023-rust/src/day11.rs

```diff
@@ -14,65 +22,128 @@ fn main() {
+    //         total_dist += src.0.abs_diff(dest.0) + src.1.abs_diff(dest.1);
+    //     }
+    // }
+    // println!("  {:?}", distances);
+    // println!("  {:?}", distances.values());
+
+    let total_dist = galaxies
+        .iter()
+        .enumerate()
+        .flat_map(|(i, &src)| galaxies
+            .iter()
+            .skip(i + 1)
+            .map(move |&dest| src.0.abs_diff(dest.0) + src.1.abs_diff(dest.1))
+        )
+        .sum();
+
+    return total_dist;
+}
+
+fn part2() -> usize {
+    println!("Parse image");
+    let img = parse_input();
+    println!("  {} x {}", img.len(), img[0].len());
+
+    let (empty_rows, empty_cols) = find_empty_rows_and_cols(&img);
+
+    // Find coords of galaxies
+    println!("Find coords of galaxies");
+    let galaxies = find_galaxies(&img);
 
     // Find distances between each pair of galaxies
     println!("Find distances between each pair of galaxies");
-    let mut distances: HashMap<String, usize> = HashMap::new();
+    let mut total_dist = 0;
+    let expansion_factor = 1_000_000;
     for x in 0..galaxies.len() {
         for y in x+1..galaxies.len() {
             let src = galaxies[x];
             let dest = galaxies[y];
-            let key = get_pair_key_str(&src, &dest);
 
             // Manhattan distance
-            let dist = abs_diff(src.0, dest.0) + abs_diff(src.1, dest.1);
-            distances.insert(key, dist);
+            total_dist += src.0.abs_diff(dest.0) + src.1.abs_diff(dest.1);
+            let num_empty_rows = empty_rows
+                .iter()
+                .filter(|&r| min(src.0, dest.0) < *r && *r < max(src.0, dest.0))
+                .count();
+            let num_empty_cols = empty_cols
+                .iter()
+                .filter(|&r| min(src.1, dest.1) < *r && *r < max(src.1, dest.1))
+                .count();
+
+            total_dist += (num_empty_rows + num_empty_cols) * (expansion_factor - 1);
         }
     }
     // println!("  {:?}", distances);
     // println!("  {:?}", distances.values());
 
-    println!("{}", distances.values().sum::<usize>());
+    return total_dist;
 }
 
-fn expand_img(img: Vec<Vec<char>>) -> Vec<Vec<char>> {
-    // Find indexes of empty rows...
-    let mut empty_rows = HashSet::new();
-    for (i, row) in img.iter().enumerate() {
-        if row.iter().all(|&e| e == '.') {
-            empty_rows.insert(i);
-        }
-    }
+fn find_empty_rows_and_cols(img: &Vec<Vec<char>>) -> (Vec<usize>, Vec<usize>) {
+    /*
+    Returns vectors of row indexes and column indexes (respectively) that correspond to empty
+    rows/columns. Each vector sorted in ascending order.
+    */
+
+    let empty_rows = (0..img.len())
+        .into_iter()
+        .filter(|&i| img[i].iter().all(|&e| e == '.'))
+        .collect();
+
+    let empty_cols = (0..img[0].len())
+        .into_iter()
+        .filter(|&j| img.iter().all(|row| row[j] == '.'))
+        .collect();
+        
+    return (empty_rows, empty_cols);
```

Comment:
This can be just `(empty_rows, empty_cols)` - last expression (not statement - no semicolon!) gets returned.